ask-sdk-model-runtime/index.ts

```diff
@@ -0,0 +1,480 @@
+                arrayHeader.push({
+                    key,
+                    value,
+                });
+            }
+        });
+
+        return arrayHeader;
+    }
+
+    /**
+     * function creating an AskSdk error.
+     * @param {string} errorScope
+     * @param {string} errorMessage
+     * @returns {Error}
+     */
+    private createAskSdkModelRuntimeError(errorScope : string, errorMessage : string) : Error {
+        const error = new Error(errorMessage);
+        error.name = `AskSdkModelRuntime.${errorScope} Error`;
+
+        return error;
+    }
+
+}
+
+/**
+ * Represents an interface that provides API configuration options needed by service clients.
+ * @interface ApiConfiguration
+ */
+export interface ApiConfiguration {
+    /**
+     * Configured ApiClient implementation
+     */
+    apiClient : ApiClient;
+    /**
+     * Authorization value to be used on any calls of the service client instance
+     */
+    authorizationValue : string;
+    /**
+     * Endpoint to hit by the service client instance
+     */
+    apiEndpoint : string;
+}
+
+/**
+ * Class to be used as the base class for the generated service clients.
+ */
+export abstract class BaseServiceClient {
+    private static isCodeSuccessful( responseCode : number ) : boolean {
+        return responseCode >= 200 && responseCode < 300;
+    }
+
+    private static buildUrl(
+        endpoint : string,
+        path : string,
+        queryParameters : Array<{ key : string, value : string }>,
+        pathParameters : Map<string, string>,
+    ) : string {
+        const processedEndpoint : string = endpoint.endsWith('/') ? endpoint.substr(0, endpoint.length - 1) : endpoint;
+        const pathWithParams : string = this.interpolateParams(path, pathParameters);
+        const isConstantQueryPresent : boolean = pathWithParams.includes('?');
+        const queryString : string = this.buildQueryString(queryParameters, isConstantQueryPresent);
+
+        return processedEndpoint + pathWithParams + queryString;
+    }
+
+    private static interpolateParams(path : string, params : Map<string, string>) : string {
+        if (!params) {
+            return path;
+        }
+
+        let result : string = path;
+
+        params.forEach((paramValue : string, paramName : string) => {
+            result = result.replace('{' + paramName + '}', encodeURIComponent(paramValue));
+        });
+
+        return result;
+    }
+
+    private static buildQueryString(params : Array<{ key : string, value : string }>, isQueryStart : boolean) : string {
+        if (!params) {
+            return '';
+        }
+
+        const sb : string[] = [];
+
+        if (isQueryStart) {
+            sb.push('&');
+        } else {
+            sb.push('?');
+        }
+
+        params.forEach((obj) => {
+            sb.push(encodeURIComponent(obj.key));
+            sb.push('=');
+            sb.push(encodeURIComponent(obj.value));

```

Comment:
Will this work with multiple query values for the same query parameter?

Reply:
Yes!